src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/RenameTypeService.cs

```diff
@@ -0,0 +1,163 @@
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    [Export(typeof(IRenameTypeService))]
+    internal class RenameTypeService : IRenameTypeService
+    {
+        private readonly VisualStudioWorkspace _workspace;
+
+        [ImportingConstructor]
+        public RenameTypeService(
+            UnconfiguredProject project,
+            VisualStudioWorkspace visualStudioWorkspace)
+        {
+            _workspace = visualStudioWorkspace;
+            SyntaxFactsServicesImpl = new OrderPrecedenceImportCollection<ISyntaxFactsService>(projectCapabilityCheckProvider: project);
+        }
+
+        [ImportMany]
+        protected OrderPrecedenceImportCollection<ISyntaxFactsService> SyntaxFactsServicesImpl { get; }
+
+        private ISyntaxFactsService SyntaxFactsService => SyntaxFactsServicesImpl.FirstOrDefault()?.Value;
+
+        public async Task<bool> AnyTypeToRenameAsync(string oldFilePath, string newFilePath, string projectPath)
+        {
+            string oldName = Path.GetFileNameWithoutExtension(oldFilePath);
+            string newName = Path.GetFileNameWithoutExtension(newFilePath);
+
+            if (!CanHandleRename(oldName, newName))
+                return false;
+
+            ISymbol symbol = await TryGetSymbolToRenameAsync(oldName, newFilePath);
+            return symbol != null;
+        }
+
+        public bool RenameType(string oldFilePath, string newFilePath, string projectPath, CancellationToken cancellationToken)
+        {
+            string oldName = Path.GetFileNameWithoutExtension(oldFilePath);
+            string newName = Path.GetFileNameWithoutExtension(newFilePath);
+
+            // NOTE: It is safe to grab the first project we encounter which
+            // contains the given file. Roslyn will handle the case where the
+            // file is included in multiple projects (linked-files or multi-TFM)
+            DocumentId documentId = _workspace.CurrentSolution.GetDocumentIdsWithFilePath(newFilePath).FirstOrDefault();
+            if (documentId is null)
+                return false;
+
+            EnvDTE.FileCodeModel codeModel = _workspace.GetFileCodeModel(documentId);
+
+            if (!TryGetCodeElementToRename(oldName, codeModel.CodeElements, out EnvDTE80.CodeElement2 codeElementToRename))
+                return false;
+
+            codeElementToRename.RenameSymbol(newName);
+            return true;
+        }
+
+        private bool CanHandleRename(string oldName, string newName)
+            => IsValidIdentifier(oldName) &&
+               IsValidIdentifier(newName) &&
+               AreNotSameIdentifierName(oldName, newName);
+
+        private bool IsValidIdentifier(string identifierName)
+            => SyntaxFactsService?.IsValidIdentifier(identifierName) ?? false;
+
+        private bool AreNotSameIdentifierName(string oldName, string newName)
+            => (SyntaxFactsService?.StringComparer.Compare(oldName, newName) != 0) == true;
```

Comment:
Why dealing with a null SyntaxFactsService? This class isn't going to work in general; is it better to just bail early? (It's particularly strange that in this case strings are neither equal nor not equal with the function further down...)